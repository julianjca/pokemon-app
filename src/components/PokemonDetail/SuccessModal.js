import React, { useState, useCallback } from "react";

import Heading from "../Heading";
import Input from "../Input";
import { Spacing } from "../Shared";
import Button from "../Button";
import { usePokemonState } from "../../context";
import { ErrorMessage } from "./styles";

const SuccessModal = ({ onClick, pokemonName }) => {
  const [pokemonNickname, setPokemonNickname] = useState("");
  const [errorMessage, setErrorMessage] = useState(null);
  const state = usePokemonState();

  const handleSubmit = useCallback(() => {
    const pokemonIndex = state.pokemons.findIndex(
      (pokemon) =>
        pokemon.pokemonName === pokemonName &&
        pokemonNickname === pokemon.pokemonNickname
    );

    if (pokemonIndex === -1) {
      onClick(pokemonNickname);
      setErrorMessage(null);
      return;
    }

    setErrorMessage("Nickname should be unique.");
  }, [onClick, pokemonNickname, pokemonName, state.pokemons]);

  return (
    <div>
      <Heading size="xl">Congratulations!</Heading>
      <Heading
        as="h3"
        size="md"
        style={{
          fontWeight: 500,
          marginTop: "10px",
        }}
      >
        You have successfully catched this pokemon.
      </Heading>
      <Heading
        as="p"
        size="sm"
        style={{
          marginTop: "10px",
          fontWeight: 400,
        }}
      >
        Give it a name:
      </Heading>
      <Spacing margin="20px 0 0">
        <Input
          value={pokemonNickname}
          onChange={(e) => setPokemonNickname(e.target.value)}
          fullWidth
          type="text"
        />
        {errorMessage && <ErrorMessage>{errorMessage}</ErrorMessage>}
      </Spacing>
      <Spacing margin="20px 0 0">
        <Button onClick={handleSubmit} fullWidth>
          Save!
        </Button>
      </Spacing>
    </div>
  );
};

export default SuccessModal;
