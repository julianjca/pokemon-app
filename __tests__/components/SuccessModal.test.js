/* eslint-disable testing-library/prefer-screen-queries */
import React from "react";
import { render, fireEvent } from "../../test-utils";
import SuccessModal from "../../src/components/PokemonDetail/SuccessModal";

describe("SuccessModal", () => {
  test("renders the text correctly", () => {
    const element = render(<SuccessModal />);

    expect(element).toMatchSnapshot();
  });

  test("runs submit function after filling pokemonNickname", () => {
    const onClick = jest.fn();
    const element = render(
      <SuccessModal onClick={onClick} pokemonName="bulbasaur" />
    );

    const input = element.container.querySelector("input");

    fireEvent.change(input, { target: { value: "robert" } });

    const button = element.getByText("Save!");

    fireEvent.click(button);

    expect(onClick).toHaveBeenCalledTimes(1);
    expect(onClick).toHaveBeenCalledWith("robert");
  });
});
