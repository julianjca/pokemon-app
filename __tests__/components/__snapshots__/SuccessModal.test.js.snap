// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SuccessModal renders the text correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .emotion-0 {
  font-weight: bold;
  color: inherit;
  font-size: 1.875rem;
}

@media (min-width: 720px) {
  .emotion-0 {
    font-size: 2.25rem;
  }
}

.emotion-0+.emotion-0 {
  margin-top: 15px;
}

.emotion-2 {
  font-weight: bold;
  color: inherit;
  font-size: 1.25rem;
}

.emotion-2+.emotion-2 {
  margin-top: 15px;
}

.emotion-4 {
  font-weight: bold;
  color: inherit;
  font-size: 1rem;
}

.emotion-4+.emotion-4 {
  margin-top: 15px;
}

.emotion-6 {
  margin: 20px 0 0;
}

.emotion-8 {
  height: 48px;
  padding: 0 1rem;
  border: 1px solid #193569;
  color: #193569;
  border-radius: 0.375rem;
  width: 100%;
  box-sizing: border-box;
}

.emotion-12 {
  background-color: #193569;
  border-radius: 0.5em;
  border: none;
  padding: 1em 1.2em;
  font-size: 1rem;
  font-weight: bold;
  color: #FDC903;
  cursor: pointer;
  -webkit-transition: 0.15s all ease-in-out;
  transition: 0.15s all ease-in-out;
  outline: none;
  width: 100%;
}

.emotion-12:hover {
  background-color: #152c57;
  -webkit-transition: 0.15s all ease-in-out;
  transition: 0.15s all ease-in-out;
}

<body>
    <div>
      <div>
        <h1
          class="emotion-0 emotion-1"
        >
          Congratulations!
        </h1>
        <h3
          class="emotion-2 emotion-1"
          style="font-weight: 500; margin-top: 10px;"
        >
          You have successfully catched this pokemon.
        </h3>
        <p
          class="emotion-4 emotion-1"
          style="margin-top: 10px; font-weight: 400;"
        >
          Give it a name:
        </p>
        <div
          class="emotion-6 emotion-7"
        >
          <input
            class="emotion-8 emotion-9"
            type="text"
            value=""
          />
        </div>
        <div
          class="emotion-6 emotion-7"
        >
          <button
            class="emotion-12 emotion-13"
          >
            Save!
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": .emotion-0 {
  font-weight: bold;
  color: inherit;
  font-size: 1.875rem;
}

@media (min-width: 720px) {
  .emotion-0 {
    font-size: 2.25rem;
  }
}

.emotion-0+.emotion-0 {
  margin-top: 15px;
}

.emotion-2 {
  font-weight: bold;
  color: inherit;
  font-size: 1.25rem;
}

.emotion-2+.emotion-2 {
  margin-top: 15px;
}

.emotion-4 {
  font-weight: bold;
  color: inherit;
  font-size: 1rem;
}

.emotion-4+.emotion-4 {
  margin-top: 15px;
}

.emotion-6 {
  margin: 20px 0 0;
}

.emotion-8 {
  height: 48px;
  padding: 0 1rem;
  border: 1px solid #193569;
  color: #193569;
  border-radius: 0.375rem;
  width: 100%;
  box-sizing: border-box;
}

.emotion-12 {
  background-color: #193569;
  border-radius: 0.5em;
  border: none;
  padding: 1em 1.2em;
  font-size: 1rem;
  font-weight: bold;
  color: #FDC903;
  cursor: pointer;
  -webkit-transition: 0.15s all ease-in-out;
  transition: 0.15s all ease-in-out;
  outline: none;
  width: 100%;
}

.emotion-12:hover {
  background-color: #152c57;
  -webkit-transition: 0.15s all ease-in-out;
  transition: 0.15s all ease-in-out;
}

<div>
    <div>
      <h1
        class="emotion-0 emotion-1"
      >
        Congratulations!
      </h1>
      <h3
        class="emotion-2 emotion-1"
        style="font-weight: 500; margin-top: 10px;"
      >
        You have successfully catched this pokemon.
      </h3>
      <p
        class="emotion-4 emotion-1"
        style="margin-top: 10px; font-weight: 400;"
      >
        Give it a name:
      </p>
      <div
        class="emotion-6 emotion-7"
      >
        <input
          class="emotion-8 emotion-9"
          type="text"
          value=""
        />
      </div>
      <div
        class="emotion-6 emotion-7"
      >
        <button
          class="emotion-12 emotion-13"
        >
          Save!
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
